### DISCUSSION TITLE
Granular access control

### DISCUSSION THREAD

**Discussion by maxmert** [modelcontextprotocol/.github#35] [General]
Is there a way to implement granular access control (apologies if I missed it in the documentation)? Not all users might have an access to all tables (or table rows in DB). Meaning that MCP needs to decide what information can be used to generate the answer based on user access permissions.

**Comment by dsp-ant** [comment]
MCP tools, resources and prompts can return errors. The MCP server is ultimately responsible for figuring out what permissions a user has. This means:
* For tools, provide a tool telling the model what it has access to. Return appropriate errors when the model attempts to access something it doesn't have access rights to.
* For resources: Ensure the initial return list of resources only contains accessible resources (e.g. database tables). Return errors in case a user attempts to access a resources that doesn't exist 
* For prompts: Return appropriate errors when the user attempts to access something it doesn't have access to.

**Reply by maxmert** [reply]
Thanks for the answer 🙏 Will try it out.
